docs/architecture/adr/ADR036-hsm-isolation-in-detail.md

```diff
@@ -0,0 +1,80 @@
+# ADR036: CloudHSM isolation
+
+## Status
+
+Accepted
+
+## Context
+
+Some of our GSP clusters provide a CloudHSM instance.  It is important to
+ensure that only authorised applications can use the CloudHSM.  There are
+different choices available for technical controls to prevent unauthorised
+access.
+
+This ADR is a post hoc document; it documents existing practice and writes down
+what we have already agreed and been doing for a while, but didn't capture as
+an ADR.
+
+### Network isolation
+
+Network isolation ensures that only authorised applications are permitted to
+open connections to the CloudHSM.  In a traditional AWS world, this could be
+achieved using security groups - each app would have its own security group,
+and the HSM security group would only allow connections from authorised apps.
+
+In Kubernetes, all pods share the same subnet and same set of IPs.  This means
+that the HSM security group cannot distinguish between traffic coming from
+different apps.
+
+Kubernetes provides [Network Policies][] to place controls on which pods are
+allowed to connect to which other pods.  Network Policies are similar to
+security groups, in that you can think of them as layer 4 stateful firewalls.
+
+We have also deployed [Istio][], an advanced networking tool which provides
+many features, but in the context of this ADR the main things it provides are
+further network controls at a higher level (broadly, at layer 7).
+
+[Network Policies]: https://kubernetes.io/docs/concepts/services-networking/network-policies/
+[Istio]: https://istio.io
+
+### Authentication and credential management
+
+In order to use the CloudHSM, you need to connect to it, but you also need to
+authenticate to it.  CloudHSM supports multiple users.  Each user has a
+different username and password.
+
+The authorised eIDAS apps have a username and password provided to them in a
+`Secret`.  Apps in other namespaces do not have access to this `Secret` and
+therefore do not have the required username and password to authenticate to the
+CloudHSM.
+
+## Decision
+
+We will prevent unauthorised network access to the CloudHSM by:
+
+ - setting each tenant namespace to have a `default-deny` egress
```

Comment:
> default-deny egress NetworkPolicy

This may not be true for long (discussion: https://github.com/alphagov/gsp/pull/414), due to the problems a blanket default-deny NetworkPolicy have caused. Instead we will switch to having the Istio sidecar "default-deny" all egress traffic to anything not explicitly allowed.

Reply:
yeah I've been keeping an eye on your work on this. I think we'll hold fire on this ADR until we've figured something else. I suspect the general things we'll be doing (restricting network access) will be similar but the concrete implementation will be different.

Reply:
ok so after the work in https://github.com/alphagov/gsp/pull/414 the following is the case:

* Istio will prevent routing of egress traffic unless it is explicitly whitelisted as part of the mesh (ie has a `ServiceEntry` or `VirtualService` defined)
* A `GlobalNetworkPolicy` will deny network traffic to HSM from any `Pod` in the cluster not blessed with the `talksToHsm: true` label